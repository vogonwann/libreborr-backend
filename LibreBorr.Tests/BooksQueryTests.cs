using HotChocolate;
using HotChocolate.Execution;
using LibreBorr.Web.GraphQl.Mutations;
using LibreBorr.Web.GraphQl.Queries;
using LibreBorr.Web.GraphQl.Subscriptions;
using Microsoft.Extensions.DependencyInjection;
using Snapshooter.Xunit;

namespace LibreBorr.Tests;

public class BooksQueryTests
{
    [Fact]
    public async Task SchemaChangeTest()
    {
        var schema = await new ServiceCollection()
            .AddGraphQL()
            .AddQueryType(d => d.Name("Query"))
                .AddType<BookQuery>()
                .AddType<FriendsQuery>()
            .AddMutationType(d => d.Name("Mutation"))
                .AddType<BookMutation>()
                .AddType<FriendMutation>()
            .AddSubscriptionType<BookSubscription>()
            .AddInMemorySubscriptions()
            .BuildSchemaAsync();
        
        schema.ToString().MatchSnapshot();
    }
    
    [Fact]
    public async Task GetBooksTest()
    {
        var schema = await new ServiceCollection()
            .AddGraphQL()
            .AddQueryType(d => d.Name("Query"))
            .AddType<BookQuery>()
            .AddType<FriendsQuery>()
            .AddMutationType(d => d.Name("Mutation"))
            .AddType<BookMutation>()
            .AddType<FriendMutation>()
            .AddSubscriptionType<BookSubscription>()
            .AddInMemorySubscriptions()
            .ExecuteRequestAsync(@"
                query books {
                  books {
                    id
                    title
                    description
                    tags {
                      name
                    }
                    image
                    authors
                    genres
                  }
                }
                ");
        
        schema.ToJson().MatchSnapshot();
    }
}