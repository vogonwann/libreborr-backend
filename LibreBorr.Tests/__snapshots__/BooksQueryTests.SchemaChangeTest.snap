schema {
  query: Query
  mutation: Mutation
  subscription: BookSubscription
}

type Book {
  id: Int
  title: String
  description: String
  image: [Byte!]!
  authors: String
  genres: String
  isbn: String
  year: Int
  tags: [Tag!]
}

type BookResponse {
  id: Int
  title: String
  description: String
  action: String!
  message: String!
  error: ErrorResponsePart
}

type BookSubscription {
  onBookAdded: BookResponse!
  bookUpdated: BookResponse!
  onBookDeleted: BookResponse!
}

type ErrorResponsePart {
  innerMessage: String
  message: String
  code: Int!
}

type Friend {
  id: Int!
  firstName: String!
  lastName: String!
  nickname: String
  books: [Book!]
}

type Mutation {
  createBook(bookInput: BookInput!): BookResponse
  updateBook(bookInput: BookInput!): BookResponse
  deleteBook(bookInput: BookInput!): BookResponse!
}

type Query {
  books: [Book!]
  book(id: Int!): Book
  wishlistBooks: [Book!]
  friends: [Friend!]!
}

type Tag {
  name: String!
}

input BookInput {
  id: Int
  title: String
  year: Int
  isbn: String
  description: String
  image: String
  authors: String
  genres: String
  tags: String
}

"The `Byte` scalar type represents non-fractional whole numeric values. Byte can represent values between 0 and 255."
scalar Byte
